### DISCUSSION TITLE
How to add custom token probabilities?

### DISCUSSION THREAD

**Discussion by dakinggg** [explosion/spaCy#8744] [Help: Coding & Implementations] [answered]
Previously (before spacy3), we loaded the probabilities for a custom model with some scripts copied over from spacy. In upgrading to 3, I apparently didn't do it right, because our probabilities aren't saved with the models at all. Now that I'm looking at upgrading to 3.1, I'd like to fix this. Our jsonl vocab file looks like this
```
{"lang": "en", "settings": {"oov_prob": -20.502029418945312}}
{"orth": "the", "prob": -3.0642620794985156}
{"orth": ",", "prob": -3.1135609097624126}
{"orth": ".", "prob": -3.361002540010581}
{"orth": "of", "prob": -3.403782345076145}
...
```

I assume I need to set something related to `[initialize.lookups]`, but I am a bit at a loss of how to do it. Could you point me towards the right part of the docs or tell me the easiest way to accomplish this? Thanks!

**Comment by adrianeboyd** [comment, marked as answer]
Sorry, the documentation on this is lacking.

If you want the default `lexeme_prob` tables from `spacy-lookups-data` it would look like this (in addition to other tables like `lexeme_norm`):

```
[initialize.lookups]
@misc = "spacy.LookupsDataLoader.v1"
lang = ${nlp.lang}
tables = ["lexeme_prob", "lexeme_settings"]
```

You probably a want a custom loader that loads tables from your source instead. The function should return a `Lookups` object.

If you want to load the tables in a script instead, you would add the tables `lexeme_prob` and `lexeme_settings` to `nlp.vocab.lookups` however you'd like.

I knew that we'd kept the option, but it's even more hidden than I thought: there is a hidden option in `spacy init vectors` called `--lexemes-jsonl` that will load the old JSONL lexeme table in the format along with the vectors into the vocab. If you use that, you could then use [`spacy.copy_from_base_model.v1`](https://spacy.io/api/top-level#copy_from_base_model) to copy the whole vocab into your pipeline on init. The `init vectors` option hasn't been tested a whole lot since v3 and I think using the other separate lookups tables options are clearer in the config, so I wouldn't really recommend it, but it would be the closest to the v2 setup.
